fix(ngcc): libraries using spread in object literals cannot be processed

devversion opened the pull request:
Consider a library that uses a shared constant for host bindings. e.g.

```ts
export const BASE_BINDINGS= {
  '[class.mat-themed]': '_isThemed',
}

----

@Directive({
  host: {...BASE_BINDINGS, '(click)': '...'}
})
export class Dir1 {}

@Directive({
  host: {...BASE_BINDINGS, '(click)': '...'}
})
export class Dir2 {}
```

Previously when these components were shipped as part of the
library to NPM, consumers were able to consume `Dir1` and `Dir2`.
No errors showed up.

Now with Ivy, when ngcc tries to process the library, an error
will be thrown. The error is stating that the host bindings should
be an object (which they obviously are). This happens because
TypeScript transforms the object spread to individual
`Object.assign` calls (for compatibility).

The partial evaluator used by the `@Directive` annotation handler
is unable to process this expression because there is no
integrated support for `Object.assign`. In View Engine, this was
not a problem because the `metadata.json` files from the library
were used to compute the host bindings.

To fix this, we add support for `Object.assign`. Unfortunately there
is no safe and easy way to detect if a property access resolves to
the `assign` method of the global `Object` variable. To ensure that
the partial evaluator can resolve both the "lhs" and "rhs" of the property
access, we add the needed `Object.assign` types to the ngcc program.

Fixes #34659

devversion commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts:
```diff
@@ -367,7 +367,13 @@ export class StaticInterpreter {
+      // Handles `Object.assign`. The interpreter intends to support the spread operator, but
+      // should also handle a TypeScript transpiled spread object literal. TypeScript transforms
+      // the spread operator into `Object.assign` calls when targeting anything older than ES2018.
+      if (identifier && identifier.text === 'Object' && rhs === 'assign') {
```
This is still unsafe. I'm not sure how we can reliably ensure that this `Reference` is the global `Object`. I don't want to check if the source-file that declares `Reference#node` matches the name of the fake source file we added to the program. This is specific to ngcc, and the isolated interpreter should not know about this implementation detail.

Any ideas?

devversion replied:
Improved by using a global definition that can be identified.

gkalpak reviewed (commented):
TBH, I would expect (and prefer) `Object.assign()` to be treated as a [BuiltinFn][1] and `__assign()` to be treated as a [TsHelperFn][2]. That would also ensure that fixing/improving the respective features (`BuiltinFn`/`TsHelperFn`) will apply to `assign` too.

(For example, there is work in progress to support imported helpers (e.g. `tslib.__<helper>()`). With the current implementation, this fix should be applied separately to `assign`.

[1]: https://github.com/angular/angular/blob/725148a44/packages/compiler-cli/src/ngtsc/partial_evaluator/src/result.ts#L56
[2]: https://github.com/angular/angular/blob/725148a44/packages/compiler-cli/src/ngtsc/reflection/src/host.ts#L334

devversion commented:
@gkalpak Thanks for your initial feedback. I initially was trying to use a `BuiltinFn`, but it turned out that it doesn't even get to [`accessHelper`](https://github.com/angular/angular/blob/725148a44db896f919fc0b6d7e256dbc35af3ec6/packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts#L328) which runs the built-in functions. This is because `Object` will resolve to a `DynamicValue`. This might be avoidable with the global type definition, or by introducing a new way that allows mapping identifiers to `ts.Node`'s. This might be useful for your PR that fixes inlined ts helper functions too. I chatted with Alex about this, and I'll be experiment with this.

In general, I agree that it would be good to handle `__assign` through a `TsHelperFn`. Though, for `Object.assign` I feel slightly different. I think we don't want to add support for it in the `Interpreter`, but rather only handle it in ngcc (with the reason of supporting transpiled libraries). If we'd add it to the interpreter, then I could imagine us arguing adding more functions like `Object.keys` etc. 

Also I wanted to mention that I think that most of the evaluation code for built-in functions or helper functions could be simplified, by providing a way to transform the call expression into a corresponding TS expression. e.g. `__spreadArray(a, b)` becomes `[...a, ...b]`. The evaluator will then handle it as expected without us having to worry about the implementation.

gkalpak commented:
> I initially was trying to use a `BuiltinFn`, but it turned out that it doesn't even get to `accessHelper` [...]
> This might be avoidable with the global type definition [...]

Yes, I believe so.

##
> or by introducing a new way that allows mapping identifiers to `ts.Node`'s. This might be useful for your PR that fixes inlined ts helper functions too.

Indeed. Looking forward to see what you'll come up with.

##
> I think we don't want to add support for `Object.assign` in the `Interpreter`, but rather only handle it in ngcc (with the reason of supporting transpiled libraries).

TBH, I don't see why :grin:
`Object.assign()` is not much different than other `BuiltinFn`s (such as `Array.prototype.slice()`) and if there are `Object.assign()` instances that are _not_ a result of transpilation, it would be weird to support them in ngcc, but not ngtsc.

##
> If we'd add it to the interpreter, then I could imagine us arguing adding more functions like `Object.keys` etc. 

Which is bad, because... 😕 :grin:

devversion committed:
fix(ngcc): libraries using spread in object literals cannot be processed

Consider a library that uses a shared constant for host bindings. e.g.

```ts
export const BASE_BINDINGS= {
  '[class.mat-themed]': '_isThemed',
}

----

@Directive({
  host: {...BASE_BINDINGS, '(click)': '...'}
})
export class Dir1 {}

@Directive({
  host: {...BASE_BINDINGS, '(click)': '...'}
})
export class Dir2 {}
```

Previously when these components were shipped as part of the
library to NPM, consumers were able to consume `Dir1` and `Dir2`.
No errors showed up.

Now with Ivy, when ngcc tries to process the library, an error
will be thrown. The error is stating that the host bindings should
be an object (which they obviously are). This happens because
TypeScript transforms the object spread to individual
`Object.assign` calls (for compatibility).

The partial evaluator used by the `@Directive` annotation handler
is unable to process this expression because there is no
integrated support for `Object.assign`. In View Engine, this was
not a problem because the `metadata.json` files from the library
were used to compute the host bindings.

Fixes #34659

alxhub commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts:
```diff
@@ -8,29 +8,45 @@
+    case BuiltinDeclaration.JsGlobalObject:
+      return new Map([['assign', new ObjectAssignBuiltinFn()]]);
+    default:
+      throw new Error(`Cannot resolve unknown builtin declaration ${builtinDecl}`);
```
```suggestion
      throw new Error(`Cannot resolve unknown builtin declaration ${BuiltinDeclaration[builtinDecl]}`);
```

(gives the name, rather than the enum index)

devversion replied:
Good idea. Worth mentioning: I took this from the `TsHelperFn` code 😄 I'll do the same there too.

alxhub commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts:
```diff
@@ -8,29 +8,45 @@
+  constructor(private lhs: ResolvedValueArray) { super(); }
 
-  evaluate(args: ResolvedValueArray): ResolvedValue {
+  evaluate(node: ts.Node, args: ResolvedValueArray): ResolvedValue {
```
This is a `ts.CallExpression`, right? We can type it more narrowly.

alxhub commented on packages/compiler-cli/src/ngtsc/reflection/src/host.ts:
```diff
@@ -395,6 +409,12 @@ export interface BaseDeclaration<T extends ts.Declaration = ts.Declaration> {
+   * If set, describes the type of builtin JavaScript declaration this
+   * declaration resolves to.
+   */
+  builtinDecl?: BuiltinDeclaration
```
In the ngtsc codebase, we don't use optional structure members (at least not often). Additionally, I think `Decl` as a suffix is a bit redundant given this is in a "Declaration" interface.
```suggestion
  builtin: BuiltinDeclaration|null;
```

alxhub commented on packages/compiler-cli/src/ngtsc/partial_evaluator/test/utils.ts:
```diff
@@ -16,17 +16,20 @@ import {getDeclaration, makeProgram} from '../../testing';
+        {name: absoluteFrom('/entry.ts'), contents: `${code}; const target$ = ${expr};`},
+        ...supportingFiles
+      ],
+      undefined, undefined, checkForErrors);
```
For readability, I tend to comment each primitive/unspecified value with its argument name:

```suggestion
      /* options */ undefined, /* host */ undefined, checkForErrors);
```

devversion replied:
Sure, sounds good to me 👍

alxhub commented on packages/compiler-cli/src/ngtsc/partial_evaluator/test/evaluator_spec.ts:
```diff
@@ -13,9 +13,8 @@ import {Reference} from '../../imports';
 import {FunctionDefinition, TsHelperFn, TypeScriptReflectionHost} from '../../reflection';
 import {getDeclaration, makeProgram} from '../../testing';
-import {DynamicValue} from '../src/dynamic';
+import {DynamicValue, EnumValue} from '../index';
```
Don't import from `index`, use a direct import to the file.

alxhub commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts:
```diff
@@ -8,29 +8,45 @@
+export function resolveBuiltinDeclaration(builtinDecl: BuiltinDeclaration): ResolvedValue {
+  switch (builtinDecl) {
+    case BuiltinDeclaration.JsGlobalObject:
+      return new Map([['assign', new ObjectAssignBuiltinFn()]]);
```
Rather than construct a new `Map` each time, let's have a shared global one.

devversion replied:
I had that initially, but figured that one could also argue that it unnecessarily allocates memory since it runs as a top-level module side-effect. Should be minimal though, so let's do it.

gkalpak reviewed (changes requested):
Nice! This approach turned out quite clean :tada:

devversion commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/ts_helpers.ts:
```diff
@@ -10,17 +10,29 @@ import * as ts from 'typescript';
+      return objectAssignBuiltinFn.evaluate(node, args);
     default:
-      throw new Error(`Cannot evaluate unknown helper ${helper} inline`);
+      throw new Error(`Cannot evaluate unknown helper ${TsHelperFn[helper]} inline`);
```
I did this upon a review comment. I don't think it's necessarily to print out the name, so let's just keep it like before, and do the same for `BuiltinDeclaration`.

devversion commented on packages/compiler-cli/src/ngtsc/reflection/src/host.ts:
```diff
@@ -340,6 +340,20 @@ export enum TsHelperFn {
+  /**
+   * Indicates the `__assign` function.
+   */
+  Assign,
```
I'm using the chronological order 😄 Will change!

JoostK commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -353,6 +353,25 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+    // We detect if the resolved declaration refers to the JavaScript global `Object` by checking
+    // if the declaration node is named `Object`, resolves to a type `ObjectConstructor`, and has
+    // been defined in a default library source file.
+    if (superDeclaration && superDeclaration.node !== null &&
```
In the compiler we tend to use explicit `null`/`undefined` checks, even if's more verbose.

```suggestion
    if (superDeclaration !== null && superDeclaration.node !== null &&
```

And yes, ngcc doesn't follow this convention very closely 😅

JoostK commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -353,6 +353,25 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+        ts.isTypeReferenceNode(superDeclaration.node.type) &&
+        ts.isIdentifier(superDeclaration.node.type.typeName) &&
+        superDeclaration.node.type.typeName.text === 'ObjectConstructor' &&
+        this.src.program.isSourceFileDefaultLibrary(superDeclaration.node.getSourceFile())) {
```
I agree with George here, let's put this whole thing in a function so we can have some early returns and assign intermediate variables where it makes sense.

JoostK commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -8,10 +8,10 @@
 import * as ts from 'typescript';
 
-import {ClassDeclaration, ClassMember, ClassMemberKind, Declaration, Decorator, FunctionDefinition, Parameter, TsHelperFn, isNamedVariableDeclaration, reflectObjectLiteral} from '../../../src/ngtsc/reflection';
+import {ClassDeclaration, ClassMember, ClassMemberKind, Declaration, Decorator, FunctionDefinition, isNamedVariableDeclaration, Parameter, reflectObjectLiteral, TsHelperFn,} from '../../../src/ngtsc/reflection';
```
```suggestion
import {ClassDeclaration, ClassMember, ClassMemberKind, Declaration, Decorator, FunctionDefinition, isNamedVariableDeclaration, Parameter, reflectObjectLiteral, TsHelperFn} from '../../../src/ngtsc/reflection';
```

JoostK commented on packages/compiler-cli/src/ngtsc/partial_evaluator/test/utils.ts:
```diff
@@ -16,17 +16,20 @@ import {getDeclaration, makeProgram} from '../../testing';
 
-export function makeExpression(code: string, expr: string, supportingFiles: TestFile[] = []): {
+export function makeExpression(
+    code: string, expr: string, supportingFiles: TestFile[] = [], checkForErrors?: boolean): {
```
Is this change still needed?

devversion replied:
good catch. removed

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -353,6 +353,25 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+        viaModule: null,
+        node: null,
+      };
+    }
```
I think the idea is to keep ngtsc as lightweight as possible?

devversion replied:
The idea was that we just want to support `__assign` and `Object.assign` in ngcc for backwards compatibility. Previously the spread operator worked fine because the actual spread AST got put into the `metadata.json`. So, explicitly using `Object.assign` didn't work in ngc either.

I don't feel strongly about it. I just don't feel like there is a compelling reason to put it into ngtsc. Happy to change it if you all feel strongly about it.

gkalpak replied:
Makes sense. Let's keep it in ngcc only for now.

devversion commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin.ts:
```diff
@@ -40,3 +40,27 @@ export class ArrayConcatBuiltinFn extends BuiltinFn {
+        return DynamicValue.fromDynamicInput(node, source);
+      } else if (!(source instanceof Map)) {
+        return DynamicValue.fromUnsupportedSyntax(node);
+      }
```
great idea. thx!

devversion commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -659,6 +659,8 @@ function getTsHelperFn(node: ts.NamedDeclaration): TsHelperFn|null {
       return TsHelperFn.Spread;
     case '__spreadArrays':
       return TsHelperFn.SpreadArrays;
+    case '__assign':
```
😄

devversion commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -8,10 +8,10 @@
 import {getNameText, hasNameIdentifier, stripDollarSuffix} from '../utils';
 
-import {Esm2015ReflectionHost, ParamInfo, getPropertyValueFromSymbol, isAssignment, isAssignmentStatement} from './esm2015_host';
+import {Esm2015ReflectionHost, getPropertyValueFromSymbol, isAssignment, isAssignmentStatement, ParamInfo} from './esm2015_host';
```
wouldn't clang-format complain if it would be case-sensitive? I'm changing it, but I don't think it's a good rule if it's not enforced somehow. FWIW: Webstorm ordered it that way for me.

gkalpak replied:
I am pretty sure clang-format used to put uppercase imports before lowercase ones.
Maybe it doesn't care any more 🤷‍♂

devversion commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/builtin_declaration.ts:
```diff
@@ -0,0 +1,30 @@
+      return jsGlobalObjectValue;
+    default:
+      throw new Error(
+          `Cannot resolve unknown builtin declaration ${BuiltinDeclaration[builtinDecl]}`);
```
See comment above.

devversion committed:
fixup! fix(ngcc): libraries using spread in object literals cannot be processed

Address feedback

devversion commented on packages/compiler-cli/src/ngtsc/reflection/src/host.ts:
```diff
@@ -340,6 +340,20 @@ export enum TsHelperFn {
+/**
+ * Possible declarations which are known.
+ */
+export enum BuiltinDeclaration {
```
done. agreed

devversion commented on packages/compiler-cli/ngcc/src/packages/entry_point_bundle.ts:
```diff
@@ -50,8 +50,7 @@ export function makeEntryPointBundle(
     allowJs: true,
-    maxNodeModuleJsDepth: Infinity,
-    noLib: true, rootDir, ...pathMappings
+    maxNodeModuleJsDepth: Infinity, rootDir, ...pathMappings
```
I didn't. It has been enabled in the past too, but just seems to have been disabled when the mocked file system has been introduced. Can surely do some measurements if needed.

gkalpak replied:
Fair enough. If it has been there before, it can't be too bad :grin:

@petebacondarwin, any idea why was this removed in https://github.com/angular/angular/commit/16d7dde2ad89eb4610c74da63e0e4b00afb8b577#diff-ac19368dcc489893e10fcaa6ff9436aeR48 (just to make sure there is nothing we should take into account)?

alxhub replied:
Probably because it was a fair bit of work to introduce the default library typings into the abstract filesystem.

I doubt including an extra .d.ts file or two will cause any performance issues.

devversion commented:
@gkalpak @JoostK Thanks for reviewing. Addressed the feedback.

gkalpak reviewed (approved):
LGTM! Great work, @devversion :100:

(I still wonder about `noLib` (https://github.com/angular/angular/pull/34661#discussion_r367637113), but that shouldn't block the PR from landing.)

JoostK reviewed (approved):
Excellent work!

devversion commented:
@gkalpak I can do some measurements, but I doubt that `noLib` affects performance significantly.

gkalpak commented:
I am not concerned about performance. My question (https://github.com/angular/angular/pull/34661#discussion_r367637113) was more about whether thare might any other reason to add/remove it.

devversion commented:
Fair enough. Maybe @petebacondarwin remembers why the change has been made. IIRC we already chatted about this on Slack, and it seemed like it was because of the mocked file system which did not contain the TypeScript default library files.

petebacondarwin commented:
I believe I did add it for one of two reasons:
a) it would be faster to process if we loaded in as few files into the program as possible
b) the mock file systems did not include these files

If everything works as expected with this removed, and there is no significant performance penalty, then I am happy.

alxhub commented on packages/compiler-cli/src/ngtsc/partial_evaluator/src/known_declaration.ts:
```diff
@@ -0,0 +1,29 @@
+    case KnownDeclaration.JsGlobalObject:
+      return jsGlobalObjectValue;
+    default:
+      throw new Error(`Cannot resolve known declaration. Received: ${decl}.`);
```
```suggestion
      throw new Error(`Cannot resolve known declaration. Received: ${KnownDeclaration[decl]}.`);
```
(prints the name, not the numeric value)

AndrewKushnir commented:
[Presubmit](http://test/OCL:291025335:BASE:291025371:1579730515280:5bc45ba7)

AndrewKushnir commented:
@devversion, could you please rebase this PR? Thank you.

devversion committed:
fixup! fix(ngcc): libraries using spread in object literals cannot be processed

Address final feedback

devversion commented:
@AndrewKushnir Rebased!
